[Nexters/FALCON#7] Cookie.sol implementation (KIP-17 inheritance)

tjdcks12 opened the pull request:
[Nexters/FALCON#7] Cookie.sol implementation (KIP-17 inheritance)

Commit:
[Nexters/FALCON#7] Cookie.sol implementation (KIP-17 inheritance)

seongahjo commented on truffle-config.js:
```diff
@@ -1,78 +1,25 @@
+      network_id: "1001",
+      gas: "8500000",
+      gasPrice: null,
+    }
```
로컬에서 노드를 띄울 수 있는 도커 이미지 없나??

tjdcks12 replied:
node 베이스 이미지로 ganache-cli 띄우는 형태로 가능! Dockerfile 루트에 추가해놔쓰~

`docker build -t ganache .`
`docker run -p 8545:8545 ganache`


**truffle 설정 아래와 같이** 
host: "localhost", port: "8545"

seongahjo commented on contracts/token/KIP17/KIP17Token.sol:
```diff
@@ -0,0 +1,12 @@
+contract KIP17Token is KIP17Full, KIP17Mintable, KIP17MetadataMintable, KIP17Burnable, KIP17Pausable {
+    constructor (string memory name, string memory symbol) public KIP17Full(name, symbol) {
+    }
+}
```
일관성있게 여기도 newline이 있으면 좋을듯

tjdcks12 replied:
이해를 정확히 못햇는데 10-11L 사이에 newline? 아니면 11-12L? 아니면 13L?

seongahjo replied:
13L인듯

seongahjo commented on contracts/token/cookiepang/Cookie.sol:
```diff
@@ -0,0 +1,48 @@
+    function getFirstCreator(uint256 _tokenId) view public returns (address){
+        return _cookies[_tokenId].ownerHistory[0];
+    }
+}
```
여기도

Commit:
[Nexters#FALCON-7] 소비 가능한 코인 생성 및 NFT 를 코인으로 소비

- 상대경로 import 전부 루트디렉토리 기준으로 변경
- SpendableKIP7 구현 및 기존 NFT 컨트랙에 해당 코인으로 구매하는 함수 생성

Commit:
[Nexters#FALCON-7] 로컬 블록체인 노드용 도커이미지 추가

tjdcks12 commented:
@Nexters/james-webb 
코딩 매커니즘이 생소해서 뻘짓을 많이하느라.. 문법공부나 패턴 공부에 시간을 많이 쏟고있으 ㅠ_ㅠ 양해바람!!

TC나 커밋 잘게 쪼개는건 조금 문법 익숙해진담에 하도록할께!!

tjdcks12 committed:
[Nexters/FALCON#7] 상대경로로 수정

tjdcks12 committed:
[Nexters/FALCON#7] Cookie 컨트랙 정의

- DataType
    KIP7Token public tradeCurrency;
    uint256 public mintingPriceForKlaytn;
    uint256 public mintingPriceForHammer;
    address[] public coFounderAddress;
    mapping(uint256 => CookieInfo) public cookieInfos;
    mapping(uint256 => address) public cookieOwners;
    mapping(address => uint256)public ownedCookiesCount;
    mapping(uint256 => uint256) public cookieKlaytnPrices;
    mapping(uint256 => uint256) public cookieHammerPrices;
    mapping(uint256 => bool) public hideCookies;
    mapping(uint256 => bool) public saleCookies;
    struct CookieInfo {
        uint256 tokenId;
        string content;
        string text;
        string imageUrl;
        string tag;
    }

- modifier
    modifier onlyCookieOwner(uint256 tokenId) {
        require(ownerOf(tokenId) == msg.sender, "NotCookieOwner");
        _;
    }

- Event
    event CookieCreated (address indexed from, uint256 indexed tokenId, string title, string content, string imageUrl, string tag, uint256 klaytnPrice, uint hammerPrice);
    event CookieTransacted (address indexed from, address indexed to, uint256 indexed tokenId);
    event CookieKlaytnPriceChanged(address indexed from, uint256 indexed tokenId, uint256 price);
    event CookieHammerPriceChanged(address indexed from, uint256 indexed tokenId, uint256 price);

- Function

  - 인출 (withdraw)
  - 발행 (klaytn, hammer)
  - 판매 등록 (saleCookies)
  - 숨기기 (hideCookies)
  - 가격변경 (changeKlaytnPrice, changeHammerPrice)
  - 삭제 (burnCookie)
  - 거래 (TransferFrom, safeTrasnferFrom)

tjdcks12 committed:
[Nexters/FALCON#7] 인터페이스 기능 구현

- 대부분의 메소드 기능 구현
- TODO : require를 통한 제약조건 추가 및 테스트케이스

tjdcks12 committed:
[Nexters/FALCON#7] EOF add new line
